### PR TITLE
[networking]: add BGP VPNs support

### LINKED ISSUE #2209: Neutron V2: BGP VPNs
https://docs.openstack.org/api-ref/network/v2/#bgp-vpns

resource types and CRUD:
* bgpvpns
* bgpvpns port associations
* bgpvpns router associations
* bgpvpns network associations

### PR DESCRIPTION
**Pull request by kayrus**
Fixes #2209 

Links to the line numbers/files in the OpenStack source code that support the
code in this PR:

* https://github.com/openstack/neutron-lib/blob/cbf17ddb0ee729ce88996daf29004ffa29efd270/neutron_lib/api/definitions/bgpvpn_routes_control.py
* https://github.com/openstack/neutron-lib/blob/master/neutron_lib/api/definitions/bgpvpn.py
* https://github.com/openstack/neutron-lib/blob/master/neutron_lib/api/definitions/bgpvpn_vni.py

### CODE REVIEW

**Comment by kayrus** [comment]
I don't know what can we do with the `TestBGPAgentRUD` tests timeouts.

**Review by EmilienM — APPROVED** [review]

File: openstack/networking/v2/extensions/bgpvpns/requests.go
Comment (mandre) [inline openstack/networking/v2/extensions/bgpvpns/requests.go:131]: [Doc](https://docs.openstack.org/api-ref/network/v2/#update-a-bgp-vpn) says we're expecting 201 on success, can we double check this is correct?

File: openstack/networking/v2/extensions/bgpvpns/requests.go
Comment (mandre) [inline openstack/networking/v2/extensions/bgpvpns/requests.go:274]: Funny, [this field is undocumented](https://docs.openstack.org/api-ref/network/v2/#id857).

File: openstack/networking/v2/extensions/bgpvpns/requests.go
Comment (mandre) [inline openstack/networking/v2/extensions/bgpvpns/requests.go:468]: Here too, [doc says the successful http code is 201](https://docs.openstack.org/api-ref/network/v2/#id857). Which is correct?

File: openstack/networking/v2/extensions/bgpvpns/results.go
Comment (mandre) [inline openstack/networking/v2/extensions/bgpvpns/results.go:43]: Again, [this field is undocumented](https://docs.openstack.org/api-ref/network/v2/#id834).

File: openstack/networking/v2/extensions/bgpvpns/results.go
Comment (mandre) [inline openstack/networking/v2/extensions/bgpvpns/results.go:73]: Why is it a pointer?

File: openstack/networking/v2/extensions/bgpvpns/results.go
Comment (mandre) [inline openstack/networking/v2/extensions/bgpvpns/results.go:422]: Another [undocumented field](https://docs.openstack.org/api-ref/network/v2/#show-port-association-details).

**Review by mandre — COMMENTED** [review]
A few questions, mostly around doc inconsistency which I can get an answer by looking at the code. I'll be doing that now.

File: internal/acceptance/openstack/networking/v2/extensions/bgpvpns/bgpvpns_test.go
Comment (mandre) [inline internal/acceptance/openstack/networking/v2/extensions/bgpvpns/bgpvpns_test.go:40]: The arguments should be reversed. It's `expected, actual`.

Comment (mandre) [reply internal/acceptance/openstack/networking/v2/extensions/bgpvpns/bgpvpns_test.go:40]: You have the same issue in a few places.

File: openstack/networking/v2/extensions/bgpvpns/requests.go
Comment (kayrus) [reply openstack/networking/v2/extensions/bgpvpns/requests.go:20]: It should be this way: `?fields=name&fields=id&networks=UUID1&networks=UUID2`
The response should return `name` and `id` of the requirces, which correspond to `networks=UUID1` or `networks=UUID2` filter.

File: openstack/networking/v2/extensions/bgpvpns/requests.go
Comment (kayrus) [reply openstack/networking/v2/extensions/bgpvpns/requests.go:131]: 201 is not returned. You can confirm this in functional tests.

File: openstack/networking/v2/extensions/bgpvpns/results.go
Comment (kayrus) [reply openstack/networking/v2/extensions/bgpvpns/results.go:43]: It's a read-only field. It's set to true, when rbac rules are used.

File: openstack/networking/v2/extensions/bgpvpns/results.go
Comment (kayrus) [reply openstack/networking/v2/extensions/bgpvpns/results.go:73]: Because we need to have an ability to set it to 0 on update.

File: openstack/networking/v2/extensions/bgpvpns/results.go
Comment (kayrus) [reply openstack/networking/v2/extensions/bgpvpns/results.go:422]: API returns these attributes.

File: internal/acceptance/openstack/networking/v2/extensions/bgpvpns/bgpvpns_test.go
Comment (kayrus) [reply internal/acceptance/openstack/networking/v2/extensions/bgpvpns/bgpvpns_test.go:40]: true, I always mix them up. Will fix soon.

File: openstack/networking/v2/extensions/bgpvpns/requests.go
Comment (kayrus) [reply openstack/networking/v2/extensions/bgpvpns/requests.go:274]: It's documented in PUT method. I've just checked whether `project_id` can be set as well for cases, when admins need to create a resource. I'll add this option as well. Otherwise cloud admins will get `Request forbidden: [POST https://network-3.qa-de-1.cloud.sap/v2.0/bgpvpn/bgpvpns/920b4f41-6553-4ec5-ad87-f3bf156a7650/router_associations], error message: {"NeutronError": {"type": "NotAuthorized", "message": "Not authorized.", "detail": ""}}`

**Comment by kayrus** [comment]
@mandre addresses all your comments.

**Review by mandre — APPROVED** [review]
This looks good, thanks. Let's just wait for the CI to return before merging.

File: openstack/networking/v2/extensions/bgpvpns/requests.go
Comment (kayrus) [reply openstack/networking/v2/extensions/bgpvpns/requests.go:474]: Correct is the one, which is returned by API and passes functional tests :)

**Comment by kayrus** [comment]
@mandre thanks, merging

**Pull request merged**
